### PR TITLE
HPCC-28572 Azure LogAccess plugin fetches component name

### PR DESCRIPTION
**Pull request by rpastrana**
- Updates ALA Components log mapping configuration
- Introduces concept of logColumn key configuration
- Updates default logcomponentcol name
- Generates Join logic if component col requested
- Extracts HPCC component name from K8s container name

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@wangkx please review.

**Review by wangkx — COMMENTED** [review]
@rpastrana 3 questions

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 1]
```diff
@@ -35,13 +35,15 @@ static constexpr const char * defaultHPCCLogProcIDCol      = "hpcc_log_procid";
 static constexpr const char * defaultHPCCLogThreadIDCol    = "hpcc_log_threadid";
 static constexpr const char * defaultHPCCLogMessageCol     = "hpcc_log_message";
 static constexpr const char * defaultHPCCLogJobIDCol       = "hpcc_log_jobid";
-static constexpr const char * defaultHPCCLogComponentCol   = "kubernetes_container_name";
+static constexpr const char * defaultHPCCLogComponentCol   = "hpcc_log_component";
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:38]: good question, fortunately this is the first time we use defaultHPCCLogComponentCol for azure

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 2]
```diff
@@ -381,14 +385,14 @@ AzureLogAnalyticsCurlClient::AzureLogAnalyticsCurlClient(IPropertyTree & logAcce
 void AzureLogAnalyticsCurlClient::getMinReturnColumns(StringBuffer & columns)
 {
     //timestamp, source component, message
-    columns.appendf("\n| project %s, %s, %s", m_globalIndexTimestampField.str(), m_componentsSearchColName.str(), defaultHPCCLogMessageCol);
+    columns.appendf("\n| project %s, %s, %s", m_globalIndexTimestampField.str(), defaultHPCCLogComponentCol, defaultHPCCLogMessageCol);
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:388]: m_componentsSearchColName contains the target column name in a foreign table, defaultHPCCLogComponentCol is the projected column name.

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 3]
```diff
@@ -438,6 +442,21 @@ void AzureLogAnalyticsCurlClient::searchMetaData(StringBuffer & search, const Lo
         {
             search.append("\n| project ");
             selectcols.getString(search, ",");
+
+            //determine if componentname field included or not
+            bool componentFieldFound = false;
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:447]: you're right no need for temp var

**Review by wangkx — APPROVED** [review]
@rpastrana looks fine now.

**Comment by rpastrana** [comment]
Thanks @wangkx 

@ghalliday please merge.

**Comment by ghalliday** [comment]
@rpastrana the change looks ok, but there isn't any information in the jira about the problem it is solving, so hard to tell if it solves it.
The summary line of the commit message doesn't really indicate what it is doing.
Also, what implications does this have for the efficiency (I notice this isn't checked in the checkboxes, or testing).
Is this a bug fix? or a new feature?

**Comment by rpastrana** [comment]
@ghalliday this commit implements missing functionality in the Azure Log Analytics logAccess plugin.
This functionality absolutely requires more cycles both on the client side and the back-end. This extra work can be optionally avoided by choosing to exclude the component field. This field provides the end-user the ability to filter based on the source container.

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 4]
```diff
@@ -631,6 +658,10 @@ void AzureLogAnalyticsCurlClient::populateKQLQueryString(StringBuffer & queryStr
 
         //Forced to format log structure in query until a proper log ingest rule is created
         queryIndex.set(m_globalIndexSearchPattern.str());
+
+        if(m_IncludeComponentName)
```
Comment (ghalliday) [inline system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:662]: trivial: normally have spaces after the if, same in line 682

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 5]
```diff
@@ -438,6 +442,20 @@ void AzureLogAnalyticsCurlClient::searchMetaData(StringBuffer & search, const Lo
         {
             search.append("\n| project ");
             selectcols.getString(search, ",");
+
+            //determine if componentname field included or not
+            m_IncludeComponentName = false;
+            if (!isEmptyString(m_componentsSearchColName))
+            {
+                ForEachItemIn(idx, selectcols)
```
Comment (ghalliday) [inline system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:450]: Future: This should be in a StringArray::contains() and/or find() function

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 6]
```diff
@@ -438,6 +442,20 @@ void AzureLogAnalyticsCurlClient::searchMetaData(StringBuffer & search, const Lo
         {
             search.append("\n| project ");
             selectcols.getString(search, ",");
+
```
Comment (ghalliday) [inline system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp:445]: Do the min and default branches need to reset the value of m_IncludeComponentName?

**Review by ghalliday — COMMENTED** [review]
@rpastrana I am still really struggling to understand this code for a few reasons.
- It is very configurable - which makes it very hard to read the code and understand what some of the tests are checking.  E.g. m_componentsSearchColName, that comes from the logMapKeyColAtt attributes, which is the @keyColumn attribute of the component section, which is "containerID".  Do we ever expect the contents of logMaps to be changed?
- The class seems to be a mixture of a shared client and a per-query.  Is it thread safe?  For instance it is modifying member m_IncludeComponentName - but that could be invalid if multiple requests are processed at the same time.  This may be the only example, but making the function const may help highlight any other issues.
- It looks like including the container id is the default (even for minimum).  What kind of impact does it have on performance.  Is it something that we particularly need?  If so should we be adding something extra to the logging to avoid a join every time we return any log information?

@jakesmith any thoughts on what is useful from the logs, rather than comments on the PR itself.

**Comment by jakesmith** [comment]
> @jakesmith any thoughts on what is useful from the logs, rather than comments on the PR itself.

The main field I can currently think of that would be useful above and beyond what we push as HPCC fields, is the container name (or pod name would be as useful in most cases).
If the deployment was on a shared K8s, the pod name would be unique, but container name might not be. In those cases, knowing the k8s namespace and possibly the helm name associated with the deployment would help.

The default ContainerLog (v1) container id field is not that helpful, in so far as quickly identifying what logs belong to which component/component type.
If joining was expensive [time] enough (and potentially $$ for [basic logging](https://learn.microsoft.com/en-us/azure/azure-monitor/logs/basic-logs-configure?tabs=portal-1%2Cportal-2#which-tables-support-basic-logs), then perhaps we should consider making the mapping of container -> container name/pod name available as a seperate dump (if retreiving an archive set of logs)
Azure's [ContinainerLog v2](https://learn.microsoft.com/en-us/azure/azure-monitor/reference/tables/containerlogv2) includes the Container name as part of it's schema, but I think v2 is preview only at the moment, once it graduates to GA, we should get this for free without the join.

**Review by ghalliday — APPROVED** [review]
@rpastrana please squash (and fix typo).  I'll scan once more before merging.

File: system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp [hunk 7]
```diff
@@ -382,25 +387,22 @@ AzureLogAnalyticsCurlClient::AzureLogAnalyticsCurlClient(IPropertyTree & logAcce
     }
 }
 
-void AzureLogAnalyticsCurlClient::getMinReturnColumns(StringBuffer & columns)
+void AzureLogAnalyticsCurlClient::getMinReturnColumns(StringBuffer & columns, bool & includeComponentName)
 {
-    //timestamp, source component, message
-    columns.appendf("\n| project %s, %s, %s", m_globalIndexTimestampField.str(), defaultHPCCLogComponentCol, defaultHPCCLogMessageCol);
+    includeComponentName = false;
+    //timestamp, message - Note: omponent information in default ALA format is expensive and there avoided
```
[the remark this answers is not in the document]
Comment (rpastrana) [reply system/logaccess/Azure/LogAnalytics/CurlClient/AzureLogAnalyticsCurlClient.cpp]: fixed.

**Comment by ghalliday** [comment]
@rpastrana I want to talk through how expensive this is before merging.  Will contact you tomorrow.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**